src/clients/FtpClient.cc

```diff
@@ -69,6 +69,61 @@ escapeIAC(const char *buf)
     return ret;
 }
 
+bool
+parseEPSV(Parser::Tokenizer &tz, const SBuf &tuple, uint16_t &outPort)
+{
+    try {
+        // '(' or need more data
+        tz.skipRequired("opening '('", SBuf("("));
+
+        // Peek the delimiter from the underlying buffer (tuple[1])
+        if (tuple.length() < 2)
+            throw TextException("empty delimiter in EPSV tuple", Here());
+
+        const auto delim = static_cast<unsigned char>(tuple[1]);
+
+        // validate delimiter: ASCII 33..126 and not a digit
+        if (delim < 33 || delim > 126 || xisdigit(delim))
```

Comment:
We have `CharacterSet` to define maps of valid vs invalid characters.
```suggestion
        static const CharacterSet delimChars = CharacterSet("epsv-delimiter", 33, 126) - CharacterSet::DIGIT;
        if (!delimChars[delim])
```